core/common/src/main/java/alluxio/underfs/BaseUnderFileSystem.java

```diff
@@ -231,4 +318,59 @@ public void stopSync(AlluxioURI uri) { }
   protected static String validatePath(String path) {
     return new AlluxioURI(path).toString();
   }
+
+  /**
+   * Represents a under filesystem operation.
+   */
+  private interface UfsOperation<T> {
+    /**
+     * Applies this operation.
+     *
+     * @return the result of this operation
+     */
+    T apply() throws IOException;
+  }
+
+  /**
+   * Retries the given under filesystem operation to resolve eventual consistency issue.
+   *
+   * @param op the under filesystem operation to retry
+   * @param description the description regarding the operation
+   * @return the operation result if operation succeed or returned true
+   */
+  private <T> T retry(UfsOperation<T> op, String description)
+      throws IOException {
+    RetryPolicy retryPolicy = getRetryPolicy();
+    IOException thrownException = null;
+    while (retryPolicy.attempt()) {
+      try {
+        T result  = op.apply();
+        if (result instanceof Boolean) {
+          if ((Boolean) result) {
```

Comment:
This is kind of questionable, does `false` always imply a failure?

Reply:
Changed to use `RetryOnFailure` and `RetryOnFalse`, `false` in our `BaseFileSystem` means failures like `unable to rename File/Directory`, `Fail to delete directory`.